tempesta_fw/http2/hpack_decoder.c

```diff
@@ -0,0 +1,648 @@
+							goto Bug;
+						}
+						DPRINTF("Decoded index: %"
+							PRIuPTR "\n", index);
+						goto Duplicate;
+					} else if (c & 0x40) {
+						DPUTS
+						    ("Reference with addition...");
+						state =
+						    HPack_State_Index |
+						    HPack_Flags_Add;
+						index = c & 0x3F;
+						if (index == 0x3F) {
+ Index:
+							GET_FLEXIBLE(index);
+							DPRINTF
+							    ("Decoded index: %"
+							     PRIuPTR "\n",
+							     index);
+							hp->index = index;
+							state =
+							    HPack_State_Value;
+							goto Get_Value;
+						}
+					} else if (c & 0x20) {
+						DPUTS("New window size...");
+						index = c & 0x1F;
+						if (index == 0x1F) {
+							state =
+							    HPack_State_Window;
+							GET_FLEXIBLE(index);
+						}
+						DPRINTF("Decoded window: %"
+							PRIuPTR "\n", index);
+						goto Set_Window;
+					} else {
+						DPUTS
+						    ("Reference with value...");
+						state = HPack_State_Index;
+						if (c & 0x10) {
+							DPUTS
+							    ("Transit header...");
+							state =
+							    HPack_State_Index |
+							    HPack_Flags_Transit;
+						}
+						index = c & 15;
+						if (index == 15) {
+							goto Index;
+						}
+					}
+					hp->index = index;
+					if (index) {
+						DPRINTF("Decoded index: %"
+							PRIuPTR "\n", index);
+					}
+					if (n) {
+						if (index) {
+							goto Get_Value;
+						}
+					} else {
+						state &= ~HPack_State_Mask;
+						state |=
+						    index ? HPack_State_Value :
+						    HPack_State_Name;
+						goto Incomplete;
+					}
+				}
+			case HPack_State_Name:
+				{
+					uchar c;
+
+					DPUTS("Decode header name length...");
+					if (unlikely(m == 0)) {
+						src = buffer_next(source, &m);
+					}
+					c = *src++;
+					n--;
+					m--;
+					length = c & 0x7F;
+					if (c & 0x80) {
+						DPUTS
+						    ("Huffman encoding used...");
+						state |=
+						    HPack_Flags_Huffman_Name;
+					}
+					if (unlikely(length == 0x7F)) {
+						state &= ~HPack_State_Mask;
+						state |=
+						    HPack_State_Name_Length;
+						GET_FLEXIBLE(length);
+					} else if (unlikely(length == 0)) {
+						*rc =
+						    Err_HPack_InvalidNameLength;
+						goto Bug;
+					}
+					DPRINTF("Name length: %" PRIuPTR "\n",
+						length);
+					field->name.len = length;
+					if (n >= length) {
+						goto Get_Name_Text;
+					} else {
+						state &= ~HPack_State_Mask;
+						state |= HPack_State_Name_Text;
+						goto Incomplete;
+					}
+				}
+			case HPack_State_Name_Text:
+				{
+					length = field->name.len;
+					if (unlikely(n < length)) {
+						DPUTS("Not enough data...");
+						break;
+					}
+ Get_Name_Text:
+					DPUTS("Decode header name...");
+					if (state & HPack_Flags_Huffman_Name) {
+						DPRINTF("Decode %" PRIuPTR
+							" bytes of Huffman data...\n",
+							length);
+						unsigned int hrc;
+
+						buffer_close(source, m);
+						hrc =
+						    huffman_decode_fragments
+						    (source, buffer, length);
```

Comment:
If I correctly understand, `'HTTP2Output *buffer'` is intended for storing the result of HPACK decoding process; regarding the integration with Tempesta FW code it seems that `HTTP2Input` and `HTTP2Output` buffers should be reworked in a way that one resulting buffer should remain (as descriptor of decoded output) and, possibly, incorporate with `TfwStr` descriptor.

Reply:
Yes, agree. The implementation is done as a standalone library with some interfaces: the library does some processing on some data which was prepared by some other layer and pass the result to another layer - different logic is ran against different representations of the same data. But it's more efficient to make all the layers to know about each other and run all the processing logic at once on the same data chunk.

This is ideology for QUIC design and Tempesta FW is built on the same principles (we interbreed TCP, TLS, and HTTP) and this makes me think that we can natively integrate QUIC into our design.